cdc/owner/ddl_manager.go

```diff
@@ -130,27 +132,39 @@ func (m *ddlManager) tick(
 	ctx context.Context,
 	checkpointTs model.Ts,
 	tableCheckpoint map[model.TableName]model.Ts,
-) ([]model.TableID, model.Barrier, error) {
-	barrier := model.Barrier{}
+) ([]model.TableID, model.Ts, *schedulepb.Barrier, error) {
+	minTableBarrierTs := model.Ts(0)
+	var barrier *schedulepb.Barrier
+	m.justSentDDL = nil
 
 	m.updateCheckpointTs(checkpointTs, tableCheckpoint)
 
 	currentTables, err := m.allTables(ctx)
 	if err != nil {
-		return nil, barrier, errors.Trace(err)
+		return nil, minTableBarrierTs, barrier, errors.Trace(err)
+	}
+
+	if m.executingDDL == nil {
+		m.ddlSink.emitCheckpointTs(m.checkpointTs, currentTables)
 	}
 
 	tableIDs, err := m.allPhysicalTables(ctx)
 	if err != nil {
-		return nil, barrier, errors.Trace(err)
+		return nil, minTableBarrierTs, barrier, errors.Trace(err)
 	}
 
-	m.ddlSink.emitCheckpointTs(m.checkpointTs, currentTables)
-
+	// drain all ddl jobs from ddlPuller
 	for {
 		ts, job := m.ddlPuller.PopFrontDDL()
+		// no more ddl jobs
 		if job == nil {
 			m.schema.schemaStorage.AdvanceResolvedTs(ts)
+			if m.redoDDLManager.Enabled() {
+				err := m.redoDDLManager.UpdateResolvedTs(ctx, m.ddlResolvedTs)
```

Comment:
```suggestion
				err := m.redoDDLManager.UpdateResolvedTs(ctx, ts)
```
It is safe to use the latest ts updates because all events have been sent to redo.